lit_llms/callbacks/steady_state_detection.py

```diff
@@ -0,0 +1,177 @@
+import operator
+import warnings
+from collections import defaultdict, deque
+from functools import partial
+from typing import Any, Optional
+
+import lightning
+import torch
+from lightning_utilities.core.imports import compare_version
+
+from lit_llms.callbacks.steady_state_utils import calc_total_time_per_node, chinchilla_metric_samples, is_steady_state
+
+
+class SteadyStateDetection(lightning.pytorch.callbacks.model_summary.ModelSummary):
+    def __init__(
+        self,
+        target_loss: float,
+        batch_size: Optional[int] = None,
+        num_params: Optional[int] = None,
+        rtol: float = 0.015,
+        atol: Optional[float] = None,
+        steady_state_det_mode: str = "iter_speed",
+        average: Optional[float] = None,
+        moving_average_window: int = 10,
+        stop_on_steady_state: bool = True,
+        steady_state_steps_before_stop: int = 10,
```

Comment:
@justusschock Should there be something like a max_steps?
What if steady state can't be achieved and I don't want to run for too long?

Reply:
Maybe that's not a realistic enough scenario

Reply:
I think there shouldn't. This should run in parallel with a max-steps setting of the trainer or an early-stopping callback.